### ISSUE TITLE
Indent error with the second function argument with eslint v1.3.1

### ISSUE THREAD

**Issue by enobufs** [eslint/eslint#3614]
Similar error was reported as issue #3591 which has been fixed since commit e8081 (I've confirmed), but I am still seeing the similar error with following code using rev 52c42b:

```js
'use strict';
/*eslint-disable no-console */
var Promise = require('bluebird');

function doSomething() {
    return Promise.resolve()
    .then(function () {
        var o = 1 + 2;
        return o;
    }, function (err) {
        console.log('err:', err);
        var o = 1 - 2;
        return o;
    });
}

doSomething()
.then(function (ans) {
    console.log('ans=%d', ans);
});
```
Here's the output:
```
test.js|11 col 9 error| Expected indentation of 12 space characters but found 8. (indent)
test.js|12 col 9 error| Expected indentation of 12 space characters but found 8. (indent)
test.js|13 col 9 error| Expected indentation of 12 space characters but found 8. (indent)
test.js|14 col 5 error| Expected indentation of 8 space characters but found 4. (indent)
```
My indent rule is `[2, 4]`

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by ilyavolodin** [comment]
It looks to me like the rule is correctly reporting missing spaces. You haven't terminated `return` statement, so it should had one extra level of indentation on the next line.
CC: @gyandeeps @BYK

**Comment by enobufs** [comment]
Promise's `then` method can take two functions. (e.g. `p.then(onFulfilled, onRejected)`), I don't really understand why the second arg was treated differently from the first. Also, this is identical to the example code in the issue #3591 in that regard.

**Comment by ilyavolodin** [comment]
Right, sorry, I wasn't paying close attention to #3591 but I would've said the same thing there, in my mind this isn't correct indentation for the first callback or second, as return statement wasn't terminated. But I'll let @gyandeeps and @BYK decide.

**Comment by enobufs** [comment]
Sorry, but what do you mean by "return statement wasn't terminated" ?

**Comment by ilyavolodin** [comment]
On line 6, `return` statement wasn't terminated (no semicolons), as such, line 7 and everything up to 14 should be indented by extra 4 spaces. I understand everyone has their own unique style of writing code, so take it as my person perspective.

**Comment by enobufs** [comment]
How does these look to your eyes?
```js
var assert = require('assert');
function it() {
    return assert.throws(function () {
        return;
    }, function (err) {
        return err.name === 'Error';
    });
}
it();
```

```js
var fluent = require('./fluent');
function doThat() {
    return fluent()
    .sort(function (a, b) {
        return a - b;
    })    
    .limit()
    .exec();
}
doThat();
```

Neither of above generates any error. Both look fairly common to me.

**Comment by icirellik** [comment]
@enobufs I have a pretty good example over here https://github.com/eslint/eslint/issues/3615 which is the same and the first example in your previous comment and does generate errors.

**Comment by targos** [comment]
Here is another example:

```js
Promise.all([
    promise1,
    promise2
]).then(function () {
    return promise3();
}).then(function () {
    // do something
}, function (e) {
    console.error(e);
});
```

```
promise.js
   9:5  error  Expected indentation of 8 space characters but found 4  indent
  10:1  error  Expected indentation of 4 space characters but found 0  indent
```

**Comment by BYK** [comment]
This is the expected behavior, introduced by #3502 to fix #3279.

Chain calls are expected to be indented unless they are multi-line such as the one @targos provided. @targos is this happening on 1.3.1 or on latest master since it shouldn't happen on master.

**Comment by targos** [comment]
@BYK  `1.3.1`. Just tested on master and it's fine, thanks !

**Comment by BYK** [comment]
@targos - thanks for the verification! :)

**Comment by enobufs** [comment]
I have confimed that @targos case (and also, @icirellik 's example) have been resolved by the commit e8081.

**BUT** My example still fails with the fix. Again, I was testing with rev 52c42b, which includes the commit e8081, and still master/HEAD even now.

Can you please test the code that I initially posted?

Only the difference is:

Mine:
```js
promise
.then(function () {
}, function (err) {
})
```
examples by @targos, @icirellik:
```js
promise.then(function () {
}, function (err) {
})
```

We have tons of fluent interface code that aligns with '.' (dot) on the same indentation as the first line.

**Comment by BYK** [comment]
@enobufs - yeah, I'm saying that is expected since chain calls need to be indented. That's why I referred to that PR and issue in my previous comment.

**Comment by targos** [comment]
@enobufs I agree with @BYK, your example should be indented from the first dot.

**Comment by enobufs** [comment]
IMHO, where the dots align should be an option... like, "indent-chain-calls" or something... By default, the first line that starts with dot should determine the indentation for the rest of the chain calls (should be zero or one indentation). (jshint/jscs have not complained about this - this being option let us migrate to eslint easier without having to disable indentation check all together)

Since the following code does not generate error with master HEAD. eslint is almost allowing both ways, though... (the error seems to only occur if the *first* chain calls takes multiple argument and the second or following argument is being a function laid across multiple-lines.)

```js
function doThis() {
    return promise
    .delay(100)  // inserting this line suppresses the errors for some reason.
    .then(function () {
        // do something
    }, function (e) {
        throw e;
    });
}
doThis();
```

"chain calls need to be indented", Did you mean, as a current limitation, or as a spec? If this is a spec, then error should occur at the beginning of the chain calls, shouldn't it?

I really wish ESlint allows both ways with optional parameter for stricter option soon.

**Comment by BYK** [comment]
>By default, the first line that starts with dot should determine the indentation for the rest of the chain calls (should be zero or one indentation).

This would prevent any errors made on the first line to be caught so I disagree.

>Since the following code does not generate error with master HEAD. eslint is almost allowing both ways, though... (the error seems to only occur if the first chain calls takes multiple argument and the second or following argument is being a function laid across multiple-lines.)

This actually looks like a bug and I can see why since I check if the first statement is a `CallExpression` or not. Since in your example it is not a `CallExpression` the rule doesn't indent.

>When you said, "chain calls need to be indented", Did you mean, as a current limitation, or as a spec? If this is a spec, then error should occur at the beginning of the chain calls, shouldn't it?

As a spec/best practice. There is no limitation around this. Agree with error should be at the start. That looks like a limitation for now though I'll look into it.

>I really wish ESlint allows both ways with optional parameter for stricter option soon.

Define stricter please. If you refer to the points I touched above, they are actually bugs and we should file them. I think this ticket can be used to discuss if we should have an option to disable this or not. What do you think?

**Comment by targos** [comment]
Perhaps it makes sense to add an option like `SwitchCase` to specify how indentation should be done in this case.

**Comment by enobufs** [comment]
@targos That would be fantastic!
I guess, something like:
```js
"indent": [2, 4, {"ChainCalls": 1}]
```
If "ChainCalls" is 0:
```js
    return promise
    .then(function () {
        // do something
    }, function (e) {
        // do other thing
    })
    .finally(function () {
            // do last thing
    });
```

If "ChainCalls" is 1:
```js
    return promise
        .then(function () {
            // do something
        }, function (e) {
            // do other thing
        })
        .finally(function () {
            // do last thing
        });
```

If "ChainCalls" is not specified (default), I would expect the first chain call would determine the indentation level, and subsequent chain calls will follow. In another word, it accepts both of above patterns.

Justification:
* jscs(jshint) and javascriptlint do not complain about indentation level for chain calls.
* Makes easier to migrate from jscs/jshint to a stricter rule with "ChainCalls".

My thoughts/wishes in ESlint:
* I hope ESlint won't dictate what to do to users.
* "Best practices" should be determined by users. Otherwise, ESlint would have to make too many assumptions into "best practices" or its rules that some users wouldn't buy.
* "Fully configurable" is truly attractive and it should be the key for ESlint to be successful.
* Please make migration from other linting tools as painless as possible.

**Comment by nzakas** [comment]
Okay, I think we've gotten a little out of the spirit of what issue discussions should be here, so let me see if I can reset.

@enobufs it would be helpful if, in the future, you can specifically say what you expected to happen and what was unexpected about what did happen. Your original description is a bit vague, and that's led to some confusion on our part. 

I'm not sure why chained calls are automatically required to be indented, but it seems rational to add an option to allow customization. `ChainedCalls` isn't the right name because not all chained calls occur on separate lines. I'd say that whatever the option is, the default should be `1` to mimic currently behavior and people can set it to `0` to disallow it.

I'd also like to ask that everyone just relax a bit. We shouldn't be imposing our own best practices on the community, and likewise, we won't always be able to accommodate every request we get from the community. For non-ESLint team members, please remember that what we're doing is difficult (especially the `indent` rule), so asking for a feature means someone is trying to figure out if it's possible and feasible. We can't support every request we get, so we need to do some digging to figure out if a request is worth the effort. (Of course, you can always help the process by trying to implement such a change yourself, sometimes that can also give clarity to all involved.)

**Comment by enobufs** [comment]
Ok, I will do my best to be clearer next time. When I posted this issue, the fix for #3591 did not fix error with my code, so I decided to file this issue. At that point, I had no clue, that the chained calls required indentation. The coding convention we established has been used across many products for a long time (hard to switch to another), and that's why strongly wished some flexibility in the rules... But, I may have sounded too demanding to you all. If so, here's my apologies. I understand the challenges and how difficult they are. I have not given up switching from jshint/jscs to eslint. Let me see how I can contribute to this project.  Best.

**Comment by BYK** [comment]
How about `MultilineChainCall`? Or simply `MultilineChain` in case we don't want to restrict this to calls, eg:

```js
a = something
    .chained
    .property;
```

**Comment by targos** [comment]
@BYK +1 for general multiline chaining

I think it should also accept this:
```js
a = something.
    chained.
    property;
```

**Comment by BYK** [comment]
@targos it would :)

**Comment by nzakas** [comment]
Hate to get technical about wording, but "chain" implies more than one, but clearly this applies to even one instance. I'd think something like `MemberExpression.multiline` would be most appropriate and follows with our existing options that use node types.

**Comment by BYK** [comment]
That works for me.
